File: frameworks/node/fastify/Implementation/routes/middleware.ts
import type { FastifyReply, FastifyRequest, HookHandlerDoneFunction } from "fastify";

import type { Routes } from "../app.ts";
import { answers, payload } from "../schemas.ts";

// rb:wiring middleware.*
/**
 * A hook is Fastify's layer in front of a handler, and a route can carry hooks of its own, so
 * these run on one route and no other. onRequest is the first point in the request's lifecycle.
 */
function noop(_request: FastifyRequest, _reply: FastifyReply, done: HookHandlerDoneFunction): void {
  done();
}

const layers = (count: number) => Array.from({ length: count }, () => noop);
// rb:end

/** middleware: no-op layers in front of the handler. */
const middleware: Routes = async (app, { payloads: p }) => {
  app.get("/middleware/none", answers(payload), async () => p.small);

  app.get("/middleware/four", { ...answers(payload), onRequest: layers(4) }, async () => p.small);

  app.get("/middleware/sixteen", { ...answers(payload), onRequest: layers(16) }, async () => p.small);
};

export default middleware;

File: frameworks/rust/warp/Implementation/routes/middleware.rs
use std::convert::Infallible;

use warp::reply::Response;
use warp::{Filter, Reply};

use crate::{Payloads, Routes};

/// middleware: no-op middleware between the route's path and its handler. warp's middleware is a
/// filter wrapped around another with wrap_fn, and these wrap the handler alone, so a request for
/// another path never runs them.
pub fn routes(p: &'static Payloads) -> Routes {
    let handler = move || warp::any().map(move || warp::reply::json(&p.small).into_response());
    // rb:handler middleware.none
    let none = warp::path!("middleware" / "none").and(warp::get()).and(handler());
    // rb:handler middleware.four
    let four_layers = warp::path!("middleware" / "four").and(warp::get()).and(four(handler()));
    // rb:handler middleware.sixteen
    let sixteen_layers = warp::path!("middleware" / "sixteen").and(warp::get()).and(four(four(four(four(handler())))));
    none.or(four_layers).unify().or(sixteen_layers).unify().boxed()
}

// rb:wiring middleware.*
/// One middleware: a filter in front of the one it wraps, which passes every request through.
fn noop(inner: impl Filter<Extract = (Response,), Error = Infallible> + Clone + Send + Sync) -> impl Filter<Extract = (Response,), Error = Infallible> + Clone + Send + Sync {
    warp::any().and(inner)
}

fn four(inner: impl Filter<Extract = (Response,), Error = Infallible> + Clone + Send + Sync) -> impl Filter<Extract = (Response,), Error = Infallible> + Clone + Send + Sync {
    inner.with(warp::wrap_fn(noop)).with(warp::wrap_fn(noop)).with(warp::wrap_fn(noop)).with(warp::wrap_fn(noop))
}
// rb:end
